### PR TITLE
add resource api

### PR DESCRIPTION
**Pull request by brauner**
This adds a resource API for a LXD server. Sending a request will currently yield cpu and memory information. Further resources for appropriate LXD objects might be added later.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
I guess the parsing functions may make more sense under lxd/util?

**Comment by brauner** [comment]
@stgraber, once I have a full implementation I move stuff around a couple of times. I just would like someone to take a look at the api structs and say if that is ok or if we'd like to do this differently. For example, right now I have a single struct `Resources` in `shared/api/resources.go` that encapsulates both a `CPU` and a `Memory` struct. Right now, unless I'm hearing objections or am seeing that this doesn't work nicely while implementing, I'd just add a "storage resource" or whatever struct to it that is only every set when a storage pool is queried. Does that make sense?

File: client/lxd_server.go [hunk 1]
```diff
@@ -53,3 +55,20 @@ func (r *ProtocolLXD) HasExtension(extension string) bool {
...
 	return false
 }
+
+// GetServerResources returns the resources available to a given LXD server
+func (r *ProtocolLXD) GetServerResources() (*api.Resources, error) {
+	if !r.HasExtension("resources") {
+		return nil, fmt.Errorf("The server is missing the required \"resources\" API extension")
+	}
+
+	resources := api.Resources{}
```
Comment (stgraber) [inline client/lxd_server.go]: I think this should be ServerResources.

File: shared/api/resource.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package api
```
Comment (stgraber) [inline shared/api/resource.go]: I think this should be moved to server.go

File: shared/api/resource.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+// Resources represents the system resources avaible for LXD
```
Comment (stgraber) [inline shared/api/resource.go]: Missing API extension comment.

File: shared/api/resource.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+type Resources struct {
+	CPU    CPU    `json:"cpu" yaml:"cpu"`
+	Memory Memory `json:"memory" yaml:"memory"`
+}
+
+// Socket represents a cpu socket on the system
```
Comment (stgraber) [inline shared/api/resource.go]: Missing API extension comment.

File: shared/api/resource.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+type Socket struct {
+	Cores          uint64 `json:"cores" yaml:"cores"`
+	Frequency      uint64 `json:"frequency" yaml:"frequency"`
+	FrequencyTurbo uint64 `json:"frequency_turbo" yaml:"frequency_turbo"`
+	Name           string `json:"name,omitempty" yaml:"name,omitempty"`
+	Vendor         string `json:"vendor,omitempty" yaml:"vendor,omitempty"`
+	Threads        uint64 `json:"threads" yaml:"threads"`
+}
+
+// CPU represents the cpu resources available on the system
```
Comment (stgraber) [inline shared/api/resource.go]: Missing API extension comment.

File: shared/api/resource.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+type CPU struct {
+	Sockets []Socket `json:"sockets" yaml:"sockets"`
+	Total   uint64   `json:"total" yaml:"total"`
+}
+
+// Memory represents the memory resources available on the system
```
Comment (stgraber) [inline shared/api/resource.go]: Missing API extension comment.

File: shared/api/resource.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package api
+
+// Resources represents the system resources avaible for LXD
+type Resources struct {
```
Comment (stgraber) [inline shared/api/resource.go]: Should be ServerResources

File: shared/api/resource.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	CPU    CPU    `json:"cpu" yaml:"cpu"`
+	Memory Memory `json:"memory" yaml:"memory"`
+}
+
+// Socket represents a cpu socket on the system
+type Socket struct {
```
Comment (stgraber) [inline shared/api/resource.go]: Should be ServerResourcesCPUSocket

File: shared/api/resource.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	Cores          uint64 `json:"cores" yaml:"cores"`
+	Frequency      uint64 `json:"frequency" yaml:"frequency"`
+	FrequencyTurbo uint64 `json:"frequency_turbo" yaml:"frequency_turbo"`
+	Name           string `json:"name,omitempty" yaml:"name,omitempty"`
+	Vendor         string `json:"vendor,omitempty" yaml:"vendor,omitempty"`
+	Threads        uint64 `json:"threads" yaml:"threads"`
+}
+
+// CPU represents the cpu resources available on the system
+type CPU struct {
```
Comment (stgraber) [inline shared/api/resource.go]: Should be ServerResourcesCPU

File: shared/api/resource.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+	Sockets []Socket `json:"sockets" yaml:"sockets"`
+	Total   uint64   `json:"total" yaml:"total"`
+}
+
+// Memory represents the memory resources available on the system
+type Memory struct {
```
Comment (stgraber) [inline shared/api/resource.go]: Should be ServerResourcesMemory

File: shared/api/resource.go (hunk 2 above)
Comment (stgraber) [reply shared/api/resource.go]: Actually, server_resources.go would be more consistent.

**Comment by stgraber** [comment]
Not sure what you mean with the storage stuff. I expected we'd add a storage_resources.go file to shared/api with a structure similar to that of server_resources.go.

Btw, we should be splitting the volumes stuff out of storage.go and move that to storage_volume.go for consistency with the other shared/api files.

File: shared/api/resource.go (hunk 2 above)
Comment (brauner) [reply shared/api/resource.go]: Right, was about to point that out if we go for `storage_resources.go`. :)

**Comment by brauner** [comment]
My original understanding was that there could be a generic "Resources" struct that could container various substructs like `cpu`, `memory`, `storage pool` and so on that each could get a `omitempty` such that they are only every filled by the various subsystem. But given that we're not implementing this as a completely separate api endpoint (like say `1.0/resources/storage` `1.0/resources/server`) splitting it out into separate - per LXD entity endpoint - it makes more sense to have separate files per entity and separate structs.

**Comment by stgraber** [comment]
Right.

**Comment by stgraber** [comment]
Sorry for changing my mind again. I think that for consistency we want:
 - shared/api/resources.go (instead of server_resources.go)
 - Call the base struct Resources instead of ServerResources

So actually pretty close to what you did except that all the nested structs must still have their name changed (e.g. ResourcesCPUSocket)

File: client/lxd_server.go (hunk 1 above)
Comment (brauner) [reply client/lxd_server.go]: obsolete

File: shared/api/resource.go (hunk 2 above)
Comment (brauner) [reply shared/api/resource.go]: obsolete

File: shared/api/resource.go (hunk 3 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: shared/api/resource.go (hunk 4 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: shared/api/resource.go (hunk 5 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: shared/api/resource.go (hunk 6 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: shared/api/resource.go (hunk 7 above)
Comment (brauner) [reply shared/api/resource.go]: obsolete

File: shared/api/resource.go (hunk 8 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: shared/api/resource.go (hunk 9 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: shared/api/resource.go (hunk 10 above)
Comment (brauner) [reply shared/api/resource.go]: done

**Comment by brauner** [comment]
@stgraber, I think we're good for another round of review!

File: lxd/resources.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
+package main
+
+import (
+	"net/http"
+
+	"github.com/gorilla/mux"
```
Comment (stgraber) [inline lxd/resources.go]: Missing separator between external and internal imports.

File: lxd/storage_utils.go [hunk 12]
```diff
@@ -319,3 +320,34 @@ func shrinkVolumeFilesystem(s storage, volumeType int, fsType string, devPath st
...
+func Statvfs(path string) (*syscall.Statfs_t, error) {
+	var st syscall.Statfs_t
+
+	err := syscall.Statfs(path, &st)
+	if err != nil {
+		return nil, err
+	}
+
+	return &st, nil
+}
```
Comment (stgraber) [inline lxd/storage_utils.go]: Feels like something that'd make more sense under util?

File: lxd/util/resources.go [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+			return nil, err
+		}
+
+		t.frequencyTurbo = uint64(freq / 1000)
+
+		threads = append(threads, t)
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
```
Comment (stgraber) [inline lxd/util/resources.go]: That's an odd error to return for something that's not db based.

os.ErrNotExist would probably be a better fit here.

File: lxd/util/resources.go [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+	return threads, nil
+}
+
+func CpuResource() (api.ResourcesCPU, error) {
+	c := api.ResourcesCPU{}
+
+	threads, err := getThreads()
+	if err != nil {
+		return c, err
```
Comment (stgraber) [inline lxd/util/resources.go]: Recently I've been trying to stay away from returning such empty structs as there's the risk that when adding another error check later, it won't be empty anymore.

How about changing this function to return a pointer and return nil on error?

File: lxd/util/resources.go [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+		cur.Name = v.name
+		cur.Vendor = v.vendor
+		cur.Frequency = v.frequency
+		cur.FrequencyTurbo = v.frequencyTurbo
+	}
+
+	return c, nil
+}
+
+func MemoryResource() (api.ResourcesMemory, error) {
```
Comment (stgraber) [inline lxd/util/resources.go]: Same comment as above wrt signature.

File: lxd/util/resources.go [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+	var buffers uint64
+	var cached uint64
+	var free uint64
+	var total uint64
+
+	mem := api.ResourcesMemory{}
+	f, err := os.Open("/proc/meminfo")
```
Comment (stgraber) [inline lxd/util/resources.go]: Fine for now, but I suspect we'll eventually want this to look at the current cgroup.

File: shared/api/resource.go (hunk 3 above)
Comment (stgraber) [reply shared/api/resource.go]: not that I can see

File: shared/api/resource.go [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+// ResourcesStoragePoolSpace represents the space available to a given storage
+// pool
+// API extension: resources
+type ResourcesStoragePoolSpace struct {
+	Used  uint64 `json:"used,omitempty" yaml:"used,omitempty"`
+	Total uint64 `json:"total" yaml:"total"`
+}
+
+// ResourcesStoragePoolInodes represents the inodes available to a given storage
+// pool
```
Comment (stgraber) [inline shared/api/resource.go]: The 3 paragraphs above have un-necessary line-wrap on their first line ("pool" being alone on the second line).

File: lxd/util/resources.go (hunk 13 above)
Comment (brauner) [reply lxd/util/resources.go]: This error previously didn't live in the db code. It has been moved there while free was redoing a bunch of stuff and so there are actually quite a lot of places in some of the storage codepaths where this needs to be replaced as well.

File: lxd/util/resources.go (hunk 14 above)
Comment (brauner) [reply lxd/util/resources.go]: Agreed. It just seemed to be our previous pattern. I'd prefer pointers as well!

File: lxd/util/resources.go (hunk 15 above)
Comment (brauner) [reply lxd/util/resources.go]: Yup.

File: lxd/util/resources.go (hunk 16 above)
Comment (brauner) [reply lxd/util/resources.go]: Because of use-cases where `LXD` runs nested?

File: lxd/storage_ceph.go [hunk 18]
```diff
@@ -2716,3 +2716,76 @@ func (s *storageCeph) StorageEntitySetQuota(volumeType int, size int64, data int
...
+		"--cluster", s.ClusterName,
+		"df")
+	if err != nil {
+		return nil, err
+	}
+
+	msg := string(buf)
+	idx := strings.Index(msg, "GLOBAL:")
+	if idx < 0 {
+		return nil, fmt.Errorf(`Output did not container expected ` +
```
Comment (brauner) [inline lxd/storage_ceph.go]: typo

File: lxd/util/resources.go (hunk 13 above)
Comment (stgraber) [reply lxd/util/resources.go]: Yeah, lets just make sure we don't add additional ones :)

File: lxd/util/resources.go (hunk 14 above)
Comment (stgraber) [reply lxd/util/resources.go]: Yeah, the lxd codebase is a bit inconsistent in that regard, the client/ package is much cleaner.

That is, it always returns pointers and takes structs as arguments when required or as pointers when optional, makes checking for nil much easier.

File: lxd/util/resources.go (hunk 16 above)
Comment (stgraber) [reply lxd/util/resources.go]: Well, nested will get you meminfo through lxcfs. I was thinking of a setup where the user limits the amount of global resources available to LXD itself through cgroups.

Hasn't really come up yet, so definitely fine for now.

File: shared/api/resource.go (hunk 3 above)
Comment (brauner) [reply shared/api/resource.go]: Thanks. done!

File: shared/api/resource.go (hunk 17 above)
Comment (brauner) [reply shared/api/resource.go]: done

File: lxd/storage_ceph.go (hunk 18 above)
Comment (brauner) [reply lxd/storage_ceph.go]: done

File: lxd/util/resources.go (hunk 14 above)
Comment (brauner) [reply lxd/util/resources.go]: done

File: lxd/util/resources.go (hunk 13 above)
Comment (brauner) [reply lxd/util/resources.go]: done

File: lxd/storage_utils.go (hunk 12 above)
Comment (brauner) [reply lxd/storage_utils.go]: moved

File: lxd/resources.go (hunk 11 above)
Comment (brauner) [reply lxd/resources.go]: added

**Comment by stgraber** [comment]
Can you add a couple of simple tests? Pretty much just checking that the --resources calls don't flat out fail would likely be good enough.

**Comment by brauner** [comment]
@stgraber, I fixed up the ceph memory parsing. If you have no objections the branch should be good to go!

**Pull request merged**